### PR TITLE
feat: raise if pivot would introduce duplicate column names

### PR DESCRIPTION
**Pull request by MarcoGorelli**
closes #14308

### CODE REVIEW

File: crates/polars-core/src/frame/mod.rs [hunk 1]
```diff
@@ -411,6 +411,29 @@ impl DataFrame {
...
+        for column in &columns {
+            let name = column.name();
+            if !names.insert(name) {
+                // If insertion fails, it means the element is already in the set (duplicate)
+                polars_bail!(duplicate = name)
+            }
+        }
+        // we drop early as the brchk thinks the &str borrows are used when calling the drop
+        // of both `columns` and `names`
+        drop(names);
```
Comment (MarcoGorelli) [inline crates/polars-core/src/frame/mod.rs:433]: taken from 

https://github.com/pola-rs/polars/blob/32df09e57818861e0758a307883b8602179bd18d/crates/polars-core/src/frame/mod.rs#L264-L266

I wouldn't have thought of this

File: crates/polars-core/src/frame/mod.rs [hunk 2]
```diff
@@ -411,6 +411,29 @@ impl DataFrame {
         DataFrame { columns }
     }
 
+    /// Create a new `DataFrame` but does not check the length of the `Series`,
+    /// only check for duplicates.
+    ///
+    /// It is advised to use [Series::new](Series::new) in favor of this method.
```
Comment (stinodego) [inline crates/polars-core/src/frame/mod.rs:417]: I think this should reference DataFrame?

```suggestion
    /// It is advised to use [DataFrame::new](DataFrame::new) in favor of this method.
```

File: crates/polars-core/src/frame/mod.rs [hunk 3]
```diff
@@ -411,6 +411,29 @@ impl DataFrame {
...
+    /// It is the callers responsibility to uphold the contract of all `Series`
+    /// having an equal length, if not this may panic down the line.
+    pub fn new_no_length_checks(columns: Vec<Series>) -> PolarsResult<DataFrame> {
+        let mut names = PlHashSet::with_capacity(columns.len());
+        for column in &columns {
+            let name = column.name();
+            if !names.insert(name) {
+                // If insertion fails, it means the element is already in the set (duplicate)
+                polars_bail!(duplicate = name)
+            }
```
Comment (stinodego) [inline crates/polars-core/src/frame/mod.rs:429]: This is a better implementation than the one in `new` - perhaps update there too? The insertion doesn't technically 'fail' though, so I would reword that comment (or just leave it out - the code is clear enough imo).

**Review by stinodego — CHANGES_REQUESTED** [review]
Looks good to me, just some small comments.

**Comment by MarcoGorelli** [comment]
yup, thanks for your review!

**Review by stinodego — APPROVED** [review]
Great!

**Review by stinodego — APPROVED** [review]
Great!

File: crates/polars-core/src/frame/mod.rs [hunk 4]
```diff
@@ -411,6 +409,28 @@ impl DataFrame {
...
 
+    /// Create a new `DataFrame` but does not check the length of the `Series`,
+    /// only check for duplicates.
+    ///
+    /// It is advised to use [DataFrame::new](DataFrame::new) in favor of this method.
+    ///
+    /// # Panic
+    /// It is the callers responsibility to uphold the contract of all `Series`
+    /// having an equal length, if not this may panic down the line.
+    pub fn new_no_length_checks(columns: Vec<Series>) -> PolarsResult<DataFrame> {
```
Comment (ritchie46) [inline crates/polars-core/src/frame/mod.rs:420]: This should be tagged `unsafe` as it doesn't check the invariant of `DataFrame`.
